docs/cfpromises.8
.TH cfengine 8 "Maintenance Commands"
.SH NAME
cfpromises \- promise compiler for cfengine 3
.SH SYNOPSIS
/usr/local/sbin/cfpromises [options]

.SH DESCRIPTION
.B cfengine
is a language based system specifically designed
for testing and configuring unix\-like systems attached to
a TCP/IP network. 
.PP
Cfengine policy is defined as a number of promises about
the basic resources in a computer. Each resource promises
to uphold certain properties described in the configuration.
.B cfagent
then performs host integrity and security checks as well
as installing and repairing system configuration.

.PP
The present version of this program is only a training version.
The full cfengine 3 deployment is estimated for the end of 2008.

.SH COMMAND LINE OPTIONS
.IP "-f (--file)"
Parse filename after this switch. By default 
.B cfagent
looks for a file called promises.cf in the working directory (usually
/var/cfengine/inputs).
.IP "-h (--help )"
Help information. Display version banner and options summary.
.SH AUTHOR
Mark Burgess, Oslo University College
.SH INFORMATION
http://www.cfengine.org



ChangeLog
3.1.5   
	suspiciousnames implemented

3.1.4	(Bugfix and Stability release)
	
	Some urgent patches to 3.1.3.
	Class validation parse bug fixed.
	Global zone handling error for solaris fixed.
	Package architectures handled correctly (bug #456).
	Reading and writing of key name "root-.pub" eliminated (bug #442, #453).
	cf-serverd crash because of race condition on SERVER_KEYSEEN fixed.
	Lock purging to avoid remnant complexity explosion (bug #430).
	Some copyright notices added that got lost.

3.1.3   (Stability release)

	Major memory leaks in cf-monitord, cf-execd, cf-serverd fixed (bug #427).
	The daemons now show no growth even with very complex policies.

	cf-serverd crash due to race condition in DeleteScope() fixed (bug #406).

	Added 30 second timeout on recv() on Linux.

	package_noverify_returncode implemented (bug #256).

	A flexible mechanism for setting classes based on return codes of
	commands has been introduced. Allows for setting promise kept,
	repaired or failed based on any return codes. This is currently
	implemented for commands-promises, package-manager commands and
	transformer in files. In classes body, see attributes
	kept_returncodes, repaired_returncodes, failed_returncodes (bug
	#248, #329).
	
	New function ip2host - reverse DNS lookup (bug #146).

3.1.2   (Scalability/efficiency release)

	Big efficiency improvements by caching output from
	cf-promises. Can also be used for much more efficent policy
	deployment (only pull if changed).

	Caching state of ps command for greater efficiency. Reloaded for each bundle.

	Index class lookup improves efficiency of class evaluation for huge configurations.
	
        Fixed issue where certain promiser strings got corrupted.

	Minor memory access issues fixed.

	Iterator bug introduced in 3.1.0 fixed

3.1.1   (Bugfix release)

	Memory leaks in server tracked down and fixed.
	List expansion bug (one list items not executed) fixed.
	Security issue introduced by change of runcommand shell policy fixed. If users defined a runcommand for cf-runagent/cf-serverd communication, possible to execute commands.
	cf-key -s command for showing key hash/IP address identity pairs

3.1.0
	Change in storage of public keys. Cfengine now hashes the public key and uses this
	as the keyname. Keys will be converted automatically.

	The old dynamic addresses lists are deprecated.
	Caching of dns and key information for greater server speed.
	Change in last-seen format reflects the public key usage.

	New package policy addupdate - installs package if not there and
	updates it otherwise.

	Support for package_changes => "bulk" in file repository as well.

	New special function readstringarrayidx, similar to readstringarray,
	but uses integer indeces. Very useful if first row elements are
	not good identifiers (e.g. contains spaces, non-unique, etc.).

	Change two log formats to use time() instead of date()
	- filechanges
	- total compliance

	Change from using md5 to sha256 as default digest for commercial version,
	community retains md5 for compat.

	Commands not returning 0 in commands-promises are flagged
	as repair_failed.

	Adjustable timeout on connect(). Defaults to 10 seconds, adjustable 
	with default_timeout in agent control.

	Redesign of the knowledge map infrastructure.

	Now possible to use variables to call methods, e.g

	methods:

	  "name $(list)" usebundle => $(list)("abc");	

	See reference manual notes

	Changes to normal ordering to optimize execution.

	Increased stability by always initializing Attribute and Promise
	structures.

	When running cf-promises in dry-run mode (-n), the user does not need
	to put binaries in WORKDIR/bin. For example, non-privileged users can verify root
	policies.

	Source control revision added in version string if run in verbose mode
	(e.g. "cf-promises -vV"). This needs some refining, uses revision of a header now.

	New semantics in return values of list functions. Null values are now allowed
	and there is no iteration over empty lists. The value "cf_null" is reserved for
	use as a null iterator.
	
3.0.5p1
        Showing paths allowed/denied access to when cf-serverd is run in verbose mode.
	Bug in server fixed for dynamic addresses.
	File handle closure bugfix - too many open databases.
	Seg fault in mount files fix.
	Twin used in cf-execd without checking.
	Check_root set wrong directory permissions at source not destination.
	Error message degraded in body definition.
	Undefined body not warned as error.
	Various build enahncements.
	Package_list_update called only once per manager, and fixed crash.
	Version number bug in packages.

3.0.5
        Encryption problems fixed - client key buffer was uninitialized.

        Classes-promisers are now automatically canonified when class
	strings are defined, to simplifying the use of variables in classes.

        New scalars sys.cf_version and sys.nova_version that hold Cfengine version information.

        Attribute package_delete_convention added, to allow customizable
	package name in delete command during update.

	package_list_update_ifelapsed limit added.
	
	Private variable $(firstrepo) is available in package_name_convention
        and package_delete_convention in order to expand the full path to 
	a package, which is required by some managers.

	Some of the threading code is rewritten and made more robust. This includes
     	synchronizing access to the lastseen database from the server.

        Bad initialization of BSD flags fixed
	Multiple variable expansion issues in control fixed for server and agent
	Allow ignore_missing_bundles to affect methods: bundles too
	Run agent trust dialogue fixed

	Bug in CPU monitoring, increasing time scale caused linear decay
	of CPU measurement.

	Bug in Setuid log storage, fix.
	
	Hooks added for new Nova virtualization promises.

	Multithreading mutex failed to collide during cfservd leading to dropped authentication under heavy load.
	
	
3.0.4

src/cfpromises.c
/* 

   Copyright (C) Cfengine AS

   This file is part of Cfengine 3 - written and maintained by Cfengine AS.
 
   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; version 3.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License  
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

  To the extent this program is licensed as part of the Enterprise
  versions of Cfengine, the applicable Commerical Open Source License
  (COSL) may apply to this file if you as a licensee so wish it. See
  included file COSL.txt.
*/

/*******************************************************************/
/*                                                                 */
/*  Promises cfpromises.c                                          */
/*                                                                 */
/*  Cfengine AS 1994/96                                           */
/*                                                                 */
/*******************************************************************/

#include "cf3.defs.h"
#include "cf3.extern.h"

/*******************************************************************/

int main (int argc,char *argv[]);

/*******************************************************************/
/* Command line options                                            */
/*******************************************************************/

 char *ID = "The promise agent is a validator and analysis tool for\n"
            "configuration files belonging to any of the components\n"
            "of Cfengine. Configurations that make changes must be\n"
            "approved by this validator before being executed.";
 
 struct option OPTIONS[14] =
      {
      { "help",no_argument,0,'h' },
      { "debug",optional_argument,0,'d' },
      { "verbose",no_argument,0,'v' },
      { "dry-run",no_argument,0,'n'},
      { "version",no_argument,0,'V' },
      { "file",required_argument,0,'f'},
      { "define",required_argument,0,'D' },
      { "negate",required_argument,0,'N' },
      { "inform",no_argument,0,'I'},
      { "diagnostic",no_argument,0,'x'},
      { "analysis",no_argument,0,'a'},
      { "reports",no_argument,0,'r'},
      { NULL,0,0,'\0' }
      };


 char *HINTS[14] =
      {
      "Print the help message",
      "Set debugging level 0,1,2,3",
      "Output verbose information about the behaviour of the agent",
      "All talk and no action mode - make no changes, only inform of promises not kept",
      "Output the version of the software",
      "Specify an alternative input file than the default",
      "Define a list of comma separated classes to be defined at the start of execution",
      "Define a list of comma separated classes to be undefined at the start of execution",
      "Print basic information about changes made to the system, i.e. promises repaired",
      "Activate internal diagnostics (developers only)",
      "Perform additional analysis of configuration",
      "Generate reports about configuration and insert into CFDB",
      NULL
      };

/*******************************************************************/
/* Level 0 : Main                                                  */
/*******************************************************************/

int main(int argc,char *argv[])

{
CheckOpts(argc,argv); 
GenericInitialize(argc,argv,"common");
ThisAgentInit();
AnalyzePromiseConflicts();
GenericDeInitialize();

if (ERRORCOUNT > 0)
   {
   CfOut(cf_verbose,""," !! Inputs are invalid\n");
   exit(1);
   }
else
   {
   CfOut(cf_verbose,""," -> Inputs are valid\n");
   exit(0);
   } 
}

/*******************************************************************/
/* Level 1                                                         */
/*******************************************************************/

void CheckOpts(int argc,char **argv)

{ extern char *optarg;
  char arg[CF_BUFSIZE];
  struct Item *actionList;
  int optindex = 0;
  int c;
  
while ((c=getopt_long(argc,argv,"ad:vnIf:D:N:VSrxM",OPTIONS,&optindex)) != EOF)
  {
  switch ((char) c)
      {
      case 'f':
          
          if (optarg && strlen(optarg) < 5)
             {
             snprintf(arg,CF_MAXVARSIZE," -f used but argument \"%s\" incorrect",optarg);
             FatalError(arg);
             }

          strncpy(VINPUTFILE,optarg,CF_BUFSIZE-1);
          VINPUTFILE[CF_BUFSIZE-1] = '\0';
          MINUSF = true;
          break;

      case 'd': 
          NewClass("opt_debug");
          switch ((optarg==NULL) ? '3' : *optarg)
             {
             case '1':
                 D1 = true;
                 DEBUG = true;
                 break;
             case '2':
                 D2 = true;
                 DEBUG = true;
                 break;
             default:
                 DEBUG = true;
                 break;
             }
          break;
          
      case 'K': IGNORELOCK = true;
          break;
                    
      case 'D': NewClassesFromString(optarg);
          break;
          
      case 'N': NegateClassesFromString(optarg,&VNEGHEAP);
          break;
          
      case 'I': INFORM = true;
          break;
          
      case 'v': VERBOSE = true;
          break;
          
      case 'n': DONTDO = true;
          IGNORELOCK = true;
	  LOOKUP = true;
          NewClass("opt_dry_run");
          break;
          
      case 'V': Version("cf-promises");
          exit(0);
          
      case 'h': Syntax("cf-promises - cfengine's promise analyzer",OPTIONS,HINTS,ID);
          exit(0);

      case 'M': ManPage("cf-promises - cfengine's promise analyzer",OPTIONS,HINTS,ID);
          exit(0);

       case 'r':
          SHOWREPORTS = true;
          break;

      case 'x': SelfDiagnostic();
          exit(0);

      case 'a':
          printf("Self-analysis is not yet implemented.");
          exit(0);
          break;

      default:  Syntax("cf-promises - cfengine's promise analyzer",OPTIONS,HINTS,ID);
          exit(1);
          
      }
  }

if (argv[optind] != NULL)
   {
   CfOut(cf_error,"","Unexpected argument with no preceding option: %s\n",argv[optind]);
   }

Debug("Set debugging\n");
}

/*******************************************************************/

void ThisAgentInit()

{
SHOWREPORTS = false;
}


/* EOF */
